hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestDeleteContainerHandler.java

```diff
@@ -225,6 +233,180 @@ public void testDeleteNonEmptyContainerDir()
+    GenericTestUtils.waitFor(() ->
+            logCapturer.getOutput().contains("container is not empty"),
+        500,
+        5 * 2000);
+    Assert.assertTrue(!isContainerDeleted(hddsDatanodeService,
+        containerId.getId()));
+    Assert.assertTrue(containerDeleteFailedNonEmptyBefore <
+        metrics.getContainerDeleteFailedNonEmptyDir());
+
+    // Now empty the container Dir and try with a non empty block table
+    Container containerToDelete = getContainerfromDN(
+        hddsDatanodeService, containerId.getId());
+    File chunkDir = new File(containerToDelete.
+        getContainerData().getChunksPath());
+    File[] files = chunkDir.listFiles();
+    if (files != null) {
+      for (File file : files) {
+        FileUtils.delete(file);
+      }
+    }
+    command = new DeleteContainerCommand(containerId.getId(), false);
+
+    // Send the delete command. It should fail as even though block count
+    // is zero there is a lingering block on disk.
+    command.setTerm(
+        cluster.getStorageContainerManager().getScmContext().getTermOfLeader());
+    nodeManager.addDatanodeCommand(datanodeDetails.getUuid(), command);
+
+
+    // Check the log for the error message when deleting non-empty containers
+    GenericTestUtils.waitFor(() ->
+            logCapturer.getOutput().
+                contains("Non-empty blocks table for container"),
+        500,
+        5 * 2000);
+    Assert.assertTrue(!isContainerDeleted(hddsDatanodeService,
+        containerId.getId()));
+    Assert.assertEquals(1,
+        metrics.getContainerDeleteFailedNonEmptyBlockDB());
+    // Send the delete command. It should pass with force flag.
+    long beforeForceCount = metrics.getContainerForceDelete();
+    command = new DeleteContainerCommand(containerId.getId(), true);
+
+    command.setTerm(
+        cluster.getStorageContainerManager().getScmContext().getTermOfLeader());
+    nodeManager.addDatanodeCommand(datanodeDetails.getUuid(), command);
+
+    GenericTestUtils.waitFor(() ->
+            isContainerDeleted(hddsDatanodeService, containerId.getId()),
+        500, 5 * 1000);
+    Assert.assertTrue(isContainerDeleted(hddsDatanodeService,
+        containerId.getId()));
+    Assert.assertTrue(beforeForceCount <
+        metrics.getContainerForceDelete());
+  }
+
+  private void clearBlocksTable(Container container) throws IOException {
+    try (DBHandle dbHandle
+             = BlockUtils.getDB(
+        (KeyValueContainerData) container.getContainerData(),
+        conf)) {
+      List<? extends Table.KeyValue<String, BlockData>>
+          blocks = dbHandle.getStore().getBlockDataTable().getRangeKVs(
+          ((KeyValueContainerData) container.getContainerData()).
+              startKeyEmpty(),
+          Integer.MAX_VALUE,
+          ((KeyValueContainerData) container.getContainerData()).
+              containerPrefix(),
+          ((KeyValueContainerData) container.getContainerData()).
+              getUnprefixedKeyFilter());
```

Comment:
Same as above, I think we can use `dbHandle.getStore().getBlockIterator(container.getContainerData().getContainerID());`

Reply:
We need the key, which is more work to reconstruct from the BlockIterator.